elastic/kibana
# [App Search] DRY helper for encoding/decoding routes that can have special characters in params

## Summary

There are two instances in App Search's routing where URLs can have special characters (user generated):

1. Document details, e.g. `/app_search/engines/national-parks-demo/documents/test@#$`
2. Analytics queries, e.g. `/app_search/engines/national-parks-demo/analytics/query_detail/test@#$`

### Testing these changes

Documents testing:

- Go to http://localhost:5601/xyz/app/enterprise_search/app_search/engines/national-parks-demo/documents
- Index a new document with the following ID:
```json
[
  {
    "id": "test!@#$%^&*[]/|;:'<>~`",
    "title": "test"
  }
]
```
- Refresh the page and click on your newly created document
- Confirm that the page correctly loads:
<img width="840" alt="" src="https://user-images.githubusercontent.com/549407/106339886-d2e21300-624c-11eb-906c-59da2f8095e1.png">

Analytics testing:

- Go to http://localhost:5601/xyz/app/enterprise_search/app_search/engines/national-parks-demo/documents (or go to localhost:3002's query tester and use that, whatever works)
- Search for test!@#$%^&*[]/|;:'<>~`
- Now go to Analytics, and find your recent search query
- Click on a link to the query from either the name in a table or the eye icon
- Confirm that the page correctly loads:
<img width="906" alt="" src="https://user-images.githubusercontent.com/549407/106340198-da55ec00-624d-11eb-972d-f9e5f3fb0ba3.png">

### NOTE

There is still a bug when you reload the page on a document or query with the special character `%` in the title. React Router tries to double-decode the `%` and unfortunately breaks the page. This is an issue with the `history` library, NOT our code, and there is nothing we can do about it: https://github.com/elastic/kibana/issues/82440

### Checklist
